### DISCUSSION TITLE
Is there a way to get feedback on whether a compilation hit cache?

### DISCUSSION THREAD

**Discussion by upsj** [ccache/ccache#1107] [Questions & answers] [answered]
We are keeping an eye on compilation times of individual files using ninja's output log. Is there a way to let ccache provide feedback on whether a file was served from cache, so we can exclude those timings/use older compilation times? Any kind of output to stdout or stderr would be fine, I would like to avoid having to parse debug output explicitly, but that would of course also be possible.

EDIT: For completeness, the probably most convenient solution presented was to use CCACHE_PREFIX to add a custom compiler wrapper that is only called in case of a cache miss, and records the compilation time.

**Comment by afbjorklund** [comment, marked as answer]
The "statslog" was supposed to provide this feature, you can also look at the (global) counters in the statistics.

```console
$ ccache --show-log-stats
Cacheable calls:  1 / 1 (100.0%)
  Hits:           1 / 1 (100.0%)
    Direct:       1 / 1 (100.0%)
    Preprocessed: 0 / 1 ( 0.00%)
  Misses:         0 / 1 ( 0.00%)
Primary storage:
  Hits:           2 / 2 (100.0%)
  Misses:         0 / 2 ( 0.00%)
```

The current output options in ccache don't really parse the file name, but it is there in the log for post-processing.

```
# foo.c
direct_cache_hit
primary_storage_hit
primary_storage_hit
```

There are two hits to storage in direct mode, one for the manifest and one for the result.

Previously there was a suggestion to cache the compile times, but it is not included yet...

* https://github.com/ccache/ccache/issues/453

And even older idea was to log the command line, mostly for cache miss investigation.

----

~~What is the format of the ninja log you are using ?~~. Never mind, it is in `.ninja_log`

```
# ninja log v5
<start>	<end>	<restat>	<target>	<cmdhash>
```

**Comment by afbjorklund** [comment]
One slight issue with the current logs (the stats log and ninja log), is that difference between source and target.

i.e. ccache logs the the source file (`foo.c`) but ninja logs the target file (`foo.o`), so these need to be "translated"

I think you can use `ninja -t query` for this, but it would need a small script to clean up the logs afterwards.

It also only seems to log wall time, so it doesn't account for overload and starvation like the "rusage" times ?

**Reply by upsj** [reply]
Yes, this is definitely an imperfect benchmarking tool, but what you provided so far (I wasn't aware of the CCACHE_LOG option, though I haven't been able to get it to work so far) seems pretty close to what we need. We are only looking at eliminating the biggest runtime uncertainty, and that is ccache.

**Reply by afbjorklund** [reply]
You can get all the information from CCACHE_LOGFILE (log_file), but like you mention it also includes tons of debugging information. We were using `grep` to sift it through, but the statslog was made so that it wouldn't have to store all that output for production builds...

* https://github.com/ccache/ccache/issues/262

**Reply by afbjorklund** [reply]
Another approach would be to set CCACHE_PREFIX, like one does for `distcc`, and let the compiler wrapper store the timiings...

**Reply by afbjorklund** [reply]
Ideally I would like to have a "libccache" for use with ninja (et al), so that it could get all the cache hits and statistics internally.

**Reply by upsj** [reply]
Ooooh, the CCACHE_PREFIX solution is perfect! I didn't know that was supported in ccache. Thank you very much!
